File: app/src/main/java/me/liuyun/bjutlgn/ui/UserActivity.java
package me.liuyun.bjutlgn.ui;

import android.app.AlertDialog;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.TextInputEditText;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.style.TextAppearanceSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.CheckedTextView;
import android.widget.Spinner;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import me.liuyun.bjutlgn.R;
import me.liuyun.bjutlgn.db.UserManager;
import me.liuyun.bjutlgn.entity.User;

public class UserActivity extends AppCompatActivity {
    @BindView(R.id.rec_list) RecyclerView recyclerView;
    @BindView(R.id.toolbar) Toolbar toolbar;
    @BindView(R.id.fab) FloatingActionButton fab;
    UserAdapter adapter;
    private UserManager userManager;
    private SharedPreferences prefs;
    private int currentId;
    private int currentPackage;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_users);
        ButterKnife.bind(this);
        setSupportActionBar(toolbar);
        toolbar.setNavigationOnClickListener(v -> onBackPressed());
        fab.setOnClickListener(v -> openUserDialog(true, null));

        LinearLayoutManager llm = new LinearLayoutManager(this);
        llm.setOrientation(LinearLayoutManager.VERTICAL);
        recyclerView.setLayoutManager(llm);
        userManager = new UserManager(this);
        adapter = new UserAdapter(userManager.getAllUsers());
        recyclerView.setAdapter(adapter);
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        new ItemTouchHelper(
                new ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP | ItemTouchHelper.DOWN, ItemTouchHelper.START) {
                    public boolean onMove(RecyclerView view, RecyclerView.ViewHolder holder, RecyclerView.ViewHolder target) {
                        final int from = holder.getAdapterPosition();
                        final int to = target.getAdapterPosition();
                        final int step = from < to ? 1 : -1;
                        User first = adapter.users.get(from);
                        int previousPos = first.getPosition();
                        for (int i = from; from < to ? i < to : i > to; i += step) {
                            User next = adapter.users.get(i + step);
                            int pos = next.getPosition();
                            next.setPosition(previousPos);
                            previousPos = pos;
                            adapter.users.set(i, next);
                            userManager.updateUser(next);
                        }
                        first.setPosition(previousPos);
                        adapter.users.set(to, first);
                        userManager.updateUser(first);
                        adapter.notifyItemMoved(from, to);
                        return true;
                    }

                    public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
                        int pos = viewHolder.getAdapterPosition();
                        userManager.deleteUser(adapter.users.get(pos).getId());
                        adapter.users.remove(pos);
                        adapter.notifyItemRemoved(pos);
                    }
                }).attachToRecyclerView(recyclerView);

        prefs = PreferenceManager.getDefaultSharedPreferences(this);
        currentId = prefs.getInt("current_user", 0);
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    void openUserDialog(boolean newUser, @Nullable User currentUser) {
        Context context = UserActivity.this;
        View dialogView = LayoutInflater.from(context)
                .inflate(R.layout.user_dialog, (ViewGroup) findViewById(R.id.user_dialog));
        TextInputEditText account = ButterKnife.findById(dialogView, R.id.account);
        TextInputEditText password = ButterKnife.findById(dialogView, R.id.password);
        Spinner spinner = ButterKnife.findById(dialogView, R.id.spinner_pack);
        if (!newUser) {
            account.setText(currentUser.getAccount());
            password.setText(currentUser.getPassword());
            spinner.setSelection(currentUser.getPack());
        }
        spinner.setOnItemSelectedListener(
                new AdapterView.OnItemSelectedListener() {
                    @Override
                    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                        currentPackage = position;
                    }

                    @Override
                    public void onNothingSelected(AdapterView<?> parent) {
                    }
                }
        );
        new AlertDialog.Builder(context)
                .setTitle(R.string.pref_user)
                .setPositiveButton(R.string.button_ok, (dialog1, which) -> {
                    if (!newUser) {
                        currentUser.setAccount(account.getText().toString());
                        currentUser.setPassword(password.getText().toString());
                        currentUser.setPack(currentPackage);
                        userManager.updateUser(currentUser);
                        adapter.users.clear();
                        adapter.users.addAll(userManager.getAllUsers());
                        adapter.notifyItemChanged(currentUser.getPosition());
                    } else {
                        userManager.insertUser(account.getText().toString(), password.getText().toString(), currentPackage);
                        adapter.users.clear();
                        adapter.users.addAll(userManager.getAllUsers());
                        adapter.notifyItemInserted(adapter.users.size());
                    }
                })
                .setNegativeButton(R.string.button_cancel, (dialog2, which) -> {
                })
                .setView(dialogView)
                .show();
    }

    class UserAdapter extends RecyclerView.Adapter<UserAdapter.UserViewHolder> {
        List<User> users;

        UserAdapter(List<User> users) {
            this.users = users;
        }

        @Override
        public UserViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
            View itemView = LayoutInflater.
                    from(viewGroup.getContext()).
                    inflate(R.layout.user_card, viewGroup, false);
            return new UserViewHolder(itemView);
        }

        @Override
        public int getItemCount() {
            if (users == null) {
                return 0;
            }
            return users.size();
        }

        @Override
        public void onBindViewHolder(UserViewHolder holder, int i) {
            User user = users.get(holder.getAdapterPosition());

            holder.itemView.setOnClickListener(v -> {
                prefs.edit()
                        .putString("account", user.getAccount())
                        .putString("password", user.getPassword())
                        .putInt("current_package", user.getPack())
                        .putInt("current_user", user.getId())
                        .apply();
                UserActivity.this.finish();
            });

            SpannableStringBuilder builder = new SpannableStringBuilder()
                    .append(user.getAccount())
                    .append("\n")
                    .append(getResources().getStringArray(R.array.packages)[user.getPack()]);
            builder.setSpan(new TextAppearanceSpan(UserActivity.this, android.R.style.TextAppearance_Small),
                    user.getAccount().length() + 1, builder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            holder.userView.setText(builder);
            if (user.getId() == currentId) {
                holder.userView.setChecked(true);
            }

            holder.buttonEdit.setOnClickListener(v -> openUserDialog(false, user));
            holder.buttonDelete.setOnClickListener(v -> {
                userManager.deleteUser(user.getId());
                adapter.users.remove(holder.getAdapterPosition());
                notifyItemRemoved(holder.getAdapterPosition());
            });
        }

        class UserViewHolder extends RecyclerView.ViewHolder {
            @BindView(R.id.user) CheckedTextView userView;
            @BindView(R.id.button_edit) Button buttonEdit;
            @BindView(R.id.button_delete) Button buttonDelete;

            UserViewHolder(View view) {
                super(view);
                ButterKnife.bind(this, view);
            }
        }
    }

}

File: app/src/main/kotlin/me/liuyun/bjutlgn/ui/UserActivity.kt
package me.liuyun.bjutlgn.ui

import android.content.SharedPreferences
import android.os.Bundle
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.TextAppearanceSpan
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.edit
import androidx.lifecycle.Observer
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import me.liuyun.bjutlgn.App
import me.liuyun.bjutlgn.R
import me.liuyun.bjutlgn.databinding.ActivityUsersBinding
import me.liuyun.bjutlgn.databinding.ItemUserBinding
import me.liuyun.bjutlgn.databinding.UserDialogBinding
import me.liuyun.bjutlgn.db.UserDao
import me.liuyun.bjutlgn.entity.User

class UserActivity : AppCompatActivity() {
    private lateinit var binding: ActivityUsersBinding
    private lateinit var adapter: UserAdapter
    private lateinit var userDao: UserDao
    private lateinit var prefs: SharedPreferences
    private var currentId: Int = 0
    private var currentPackage: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityUsersBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.appBar.toolbar)
        binding.appBar.toolbar.setNavigationOnClickListener { onBackPressed() }
        binding.fab.setOnClickListener { openUserDialog(true, User(0, "", "", 0, 0)) }

        userDao = (application as App).appDatabase.userDao()
        adapter = UserAdapter()
        val users = userDao.all()
        users.observe(this@UserActivity, Observer {
            it?.let {
                val diff = DiffUtil.calculateDiff(UsersDiffCallback(adapter.users, it))
                adapter.users = it
                diff.dispatchUpdatesTo(adapter)
            }
        })

        binding.recycler.adapter = adapter
        binding.recycler.itemAnimator = DefaultItemAnimator()
        ItemTouchHelper(
                object : ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP or ItemTouchHelper.DOWN, ItemTouchHelper.START) {
                    override fun onMove(view: RecyclerView, holder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                        adapter.onItemMove(holder.adapterPosition, target.adapterPosition)
                        return true
                    }

                    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                        adapter.onItemDismiss(viewHolder.adapterPosition)
                    }
                }).attachToRecyclerView(binding.recycler)

        prefs = PreferenceManager.getDefaultSharedPreferences(this)
        currentId = prefs.getInt("current_user", 0)
    }

    internal fun openUserDialog(newUser: Boolean, user: User) {
        val binding = UserDialogBinding.inflate(layoutInflater, null, false)
        with(binding) {
            account.setText(user.account)
            account.setSelection(user.account.length)
            password.setText(user.password)

            val items = resources.getStringArray(R.array.packages)
            val adapter = ArrayAdapter(this@UserActivity, R.layout.item_package, items)
            spinnerPack.setAdapter(adapter)
            if (!newUser) {
                spinnerPack.setText(items[user.pack], false)
            }
            spinnerPack.setOnItemClickListener { _, _, position, _ ->
                currentPackage = position
            }
        }
        with(MaterialAlertDialogBuilder(this)) {
            setView(binding.root)
            setPositiveButton(R.string.button_ok) { _, _ ->
                user.account = binding.account.text.toString()
                user.password = binding.password.text.toString()
                if (!newUser) {
                    user.pack = currentPackage
                    userDao.update(user)
                } else {
                    user.position = userDao.maxPosition()?.let { it.position + 1 } ?: 0
                    userDao.insert(user)
                }
            }
            setNegativeButton(R.string.button_cancel) { _, _ -> Unit }
            show()
        }
    }

    internal inner class UserAdapter(var users: MutableList<User> = mutableListOf()) : RecyclerView.Adapter<UserAdapter.UserViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, i: Int) =
                UserViewHolder(ItemUserBinding.inflate(LayoutInflater.from(parent.context), parent, false))

        override fun getItemCount() = users.size

        override fun onBindViewHolder(holder: UserViewHolder, i: Int) {
            val user = users[holder.adapterPosition]

            holder.itemView.setOnClickListener {
                prefs.edit {
                    putString("account", user.account)
                    putString("password", user.password)
                    putInt("current_package", user.pack)
                    putInt("current_user", user.id)
                }
                this@UserActivity.finish()
            }

            val builder = SpannableStringBuilder()
                    .append(user.account)
                    .append("\n")
                    .append(resources.getStringArray(R.array.packages)[user.pack])
            builder.setSpan(TextAppearanceSpan(this@UserActivity, android.R.style.TextAppearance_Small),
                    user.account.length + 1, builder.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            holder.binding.user.text = builder
            if (user.id == currentId) {
                holder.binding.user.isChecked = true
            }

            holder.binding.buttonEdit.setOnClickListener { openUserDialog(false, user.copy()) }
            holder.binding.buttonDelete.setOnClickListener { userDao.delete(user) }
        }

        fun onItemMove(from: Int, to: Int) {
            val step = if (from < to) 1 else -1
            val first = users[from]
            var previousPos = first.position
            var i = from
            while (if (from < to) i < to else i > to) {
                val next = users[i + step]
                val pos = next.position
                next.position = previousPos
                previousPos = pos
                users[i] = next
                userDao.update(next)
                i += step
            }
            first.position = previousPos
            users[to] = first
            userDao.update(first)
            notifyItemMoved(from, to)
        }

        fun onItemDismiss(pos: Int) {
            userDao.delete(adapter.users[pos])
        }

        internal inner class UserViewHolder(val binding: ItemUserBinding) : RecyclerView.ViewHolder(binding.root)
    }

    internal class UsersDiffCallback(private val oldUsers: List<User>, private val newUsers: List<User>) : DiffUtil.Callback() {
        override fun getOldListSize() = oldUsers.size
        override fun getNewListSize() = newUsers.size
        override fun areItemsTheSame(p0: Int, p1: Int) = oldUsers[p0].id == newUsers[p1].id
        override fun areContentsTheSame(p0: Int, p1: Int) = oldUsers[p0] == newUsers[p1]
    }

}
